Mods/SML/Source/SML/Private/Patching/BlueprintHookManager.cpp

```diff
@@ -35,102 +35,172 @@ void DebugDumpFunctionScriptCode(UFunction* Function, int32 HookOffset, const FS
 #endif
 }
 
-int32 UBlueprintHookManager::PreProcessHookOffset(UFunction* Function, int32 HookOffset) {
-	if (HookOffset == EPredefinedHookOffset::Return) {
-		//For now Kismet Compiler will always generate only one Return node, so all
-		//execution paths will end up either with executing it directly or jumping to it
-		//So we need to hook only in one place to handle all possible execution paths
-		FSMLKismetBytecodeDisassembler Disassembler;
-		int32 ReturnOffset;
-		const bool bIsValid = Disassembler.FindFirstStatementOfType(Function, 0, EX_Return, ReturnOffset);
-		checkf(bIsValid, TEXT("EX_Return not found for function %s"), *Function->GetPathName());
-		return ReturnOffset;
+void UBlueprintHookManager::ModifyOffsetsForNewHookOffset(TArray<uint8>& Script, TSharedPtr<FJsonObject> Expression, int32 HookOffset)
+{
+	int32 Opcode;
+	if (!Expression->TryGetNumberField(TEXT("Opcode"), Opcode)) {
+		// No opcode means it's not an instruction, so we can just return;
+		return;
+	}
+
+	//A computed jump could do anything at runtime - it could jump to before or after the hook, so we have no way of knowing how
+	//it needs to be modified to work correctly. For now, the only predictable solution is to forbid hooking functions with them.
+	fgcheckf(Opcode != EX_ComputedJump, TEXT("Cannot hook a blueprint function that contains an EX_ComputedJump instruction. There's no way to guarantee it would not crash at some point."));
+
+	int32 IndexAfterOpcode = 1 + Expression->GetIntegerField(TEXT("OpcodeIndex"));
+
+	// This switch was created by comparing to the serialization done in FSMLKismetBytecodeDisassembler::SerializeExpression to
+	// identify which opcodes can jump to offsets and where exactly those jump targets reside in the script.
+	switch (Opcode)
+	{
+	case EX_Jump:
+	case EX_JumpIfNot:
+	case EX_Skip:
+	case EX_PushExecutionFlow:
+		{
+			int32 IndexOfCurrentJumpOffset = IndexAfterOpcode;
+			int32 CurrentJumpOffset = Expression->GetIntegerField(TEXT("Offset"));
+			if (CurrentJumpOffset > HookOffset) {
+				FPlatformMemory::WriteUnaligned(&Script[IndexOfCurrentJumpOffset], (CodeSkipSizeType)(CurrentJumpOffset + UBlueprintHookManager::JumpBytesRequired));
+			}
+			break;
+		}
+	case EX_ClassContext:
+	case EX_Context:
+	case EX_Context_FailSilent:
+		{
+			int32 CurrentJumpOffset = Expression->GetIntegerField(TEXT("SkipOffsetForNull"));
+			if (CurrentJumpOffset > HookOffset) {
+				// The offset is just past the Context object, so we have to get its size so we can write to the correct place
+				TSharedPtr<FJsonObject> Context = Expression->GetObjectField(TEXT("Context"));
+				int32 SizeOfContext = (int32)Context->GetNumberField(TEXT("OpSizeInBytes"));
+				int32 IndexOfCurrentJumpOffset = IndexAfterOpcode + SizeOfContext;
+				FPlatformMemory::WriteUnaligned(&Script[IndexOfCurrentJumpOffset], (CodeSkipSizeType)(CurrentJumpOffset + UBlueprintHookManager::JumpBytesRequired));
+			}
+			break;
+		}
```

Comment:
It turns out that in the case of Context opcodes, SkipOffsetForNull is actually an offset, not an absolute script index https://github.com/satisfactorymodding/UnrealEngine/blob/08d00ee44347101896506246689e3b779bc1bc16/Engine/Source/Runtime/CoreUObject/Private/UObject/ScriptCore.cpp#L3141C26-L3141C32

Reply:
Good catch - does this mean the absolute jump location is the OpCodeIndex + SkipOffsetForNull?  Or does the Step call at the beginning of that function mean it's OpCodeIndex + 1 + SkipOffsetForNull?

Reply:
I believe it is `Expression->GetObjectField("Expression")->GetIntegerField("OpcodeIndex") + SkipOffsetForNull`, since it reads all the data up to the `RValueProperty`, which is immediately followed by `Expression`
https://github.com/satisfactorymodding/SatisfactoryModLoader/blob/04025f8209079ee869269fa3a64932fa3cde2b4b/Mods/SML/Source/SML/Private/Toolkit/KismetBytecodeDisassembler.cpp#L480-L492
https://github.com/satisfactorymodding/UnrealEngine/blob/08d00ee44347101896506246689e3b779bc1bc16/Engine/Source/Runtime/CoreUObject/Private/UObject/ScriptCore.cpp#L3143C58-L3143C72 (the comment here seems to be wrong though, it has an extra byte skipped)

This also looks to be matching the expected index (149 + 29 = 178) in
```json
{
    "Opcode": 25,
    "OpcodeIndex": 127,
    "Inst": "Context",
    "Context":
    {
        "Opcode": 0,
        "OpcodeIndex": 128,
        "Inst": "LocalVariable",
        "VariableType":
        {
            "PinCategory": "Object",
            "PinSubCategory": "Object",
            "PinSubCategoryObject": "/Script/SML.SMLRemoteCallObject"
        },
        "VariableName": "K2Node_DynamicCast_AsSMLRemote_Call_Object",
        "OpSizeInBytes": 9
    },
    "SkipOffsetForNull": 29,
    "Expression":
    {
        "Opcode": 27,
        "OpcodeIndex": 149,
        "Inst": "VirtualFunction",
        "Function": "SendChatMessage",
        "Parameters": [
            {
                "Opcode": 31,
                "OpcodeIndex": 162,
                "Inst": "StringConst",
                "Value": "asdf",
                "OpSizeInBytes": 6
            },
            {
                "Opcode": 0,
                "OpcodeIndex": 168,
                "Inst": "LocalVariable",
                "VariableType":
                {
                    "PinCategory": "Struct",
                    "PinSubCategory": "Struct",
                    "PinSubCategoryObject": "/Script/CoreUObject.LinearColor"
                },
                "VariableName": "CallFunc_MakeColor_ReturnValue",
                "OpSizeInBytes": 9
            }
        ],
        "OpSizeInBytes": 29
    },
    "OpSizeInBytes": 51,
    "StatementIndex": 127
},
{
    "Opcode": 4,
    "OpcodeIndex": 178,
    "Inst": "Return",
    "Expression":
    {
        "Opcode": 11,
        "OpcodeIndex": 179,
        "Inst": "Nothing",
        "OpSizeInBytes": 1
    },
    "OpSizeInBytes": 2,
    "StatementIndex": 178
}
```

Reply:
I've also now noticed a few remaining uses of `GetNumberField` casted to `int32`

Reply:
Because it's a relative offset, I'm realizing that it needs to be adjusted if either:

1) The hook insertion is before the Context instruction and the final jump is backwards to AT/BEFORE the hook insertion (in which case it needs to jump 5 further backwards) OR
2) The hook insertion is after the Context instruction and the final jump is forwards to AFTER the hook insertion.

I was thinking of detecting 2 by comparing the final jump offset to the start of the statement after the Context instruction.  But I realize the Context can be an expression within a larger statement.  Do I need to find the next outermost statement or can I just detect it jumping ahead and assume it will be a statement somewhere after the Context instruction?  I think that's OK but I need to know if I'm missing anything.